Preformatted citation formats

mbollmann opened the pull request:
This continues the discussion in #1375 about having preformatting citation formats (such as an ACL-style bibliography entry, short markdown citation strings, etc.).

Summary of the state of things before splitting this off: https://github.com/acl-org/acl-anthology/pull/1375#issuecomment-867061645

akoehn commented:
Writing reviews, I now really want plain text citations, would be so helpful :-)

akoehn commented:
I think I mentioned this before in #1375: maybe we can handle this in javascript instead, e.g. with something like https://github.com/pcooksey/bibtex-js.  Then we would not need additional processing while building the anthology, it is just a different rendering of the bibtex file we already have.

nschneid commented:
Or we could implement a simple solution directly from the Anthology fields. I suspect we don't need most of the complexity of BibTeX to generate a citation given that there are only a couple of relevant item types (conference papers, journal articles—anything else? full proceedings volumes maybe?).

I suppose there are a small number of titles with LaTeX math or markup, but it might be easier to use a library for those directly than to convert the whole BibTeX entry.

mbollmann commented:
> I think I mentioned this before in #1375: maybe we can handle this in javascript instead, e.g. with something like https://github.com/pcooksey/bibtex-js. Then we would not need additional processing while building the anthology, it is just a different rendering of the bibtex file we already have.

I'm not familiar with these options, but the one you linked doesn't use .bst files; custom bib rendering works via HTML templates. In that case, we might just as well implement that rendering server-side. The big advantage of a "proper" BibTeX renderer IMHO would be offering several commonly used bibliography styles (MLA, APA, etc.).

> I suppose there are a small number of titles with LaTeX math or markup, but it might be easier to use a library for those directly than to convert the whole BibTeX entry.

Those are handled anyway for rendering the title on the website, so that's not an issue :)

mjpost commented:
Maybe your interest in finishing this off has waned, but I find myself in want of this feature all the time, basically any time I’m building a list of papers to share via email or in private docs.

mbollmann commented:
I remember this as essentially being a finished feature suggestion, with the discussion being left open whether we want to do this in JavaScript instead, and no one doing a code review/approval so far :)

Can look into the merge conflicts tomorrow so we can rebuild the preview.

mjpost commented:
Oh, great. Yes, if you can address the merge conflicts so that the preview is rebuilt, I’ll expedite a review!

nschneid commented:
> There was also the question of how to produce these citation strings; the implementation here adds another bit of overhead to the build process, which led Arne to suggest to do this in JavaScript instead (I'm not sure I'd be happy with that but I haven't tried it yet either).

I will register a general bias against doing anything in Javascript that can be done server-side.

nschneid commented:
I notice short cite URLs are showing up with `/None` at the end—is that a bug?

mjpost commented:
Can you link an example of what you mean by /None?

Agreed we should follow ACL cite format and switch to "et al." on author lists.

For most venues prior to 2020 we unfortunately do not have the real venue name. I think the quick fix is to remove the venue from the cite if it is "WS". The better undertaking is a massive effort to label all such volumes.

I would be grateful to see the cite formats have the title link to the ACL page. People can always paste as plain text to remove it.

mbollmann commented:
Yeah, in the meantime we made a change how URLs are accessed internally, so the links all currently link to "None". (can be seen best in the "Export citation" -> "Markdown" popup

nschneid commented:
> Agreed we should follow ACL cite format and switch to "et al." on author lists.

ACL format has no more than 2 authors.

mbollmann commented:
Another point: if we have no need for other citation formats than ACL and informal Markdown, we might want to consider replacing the `citeproc` dependency I'm currently using with a small Python class that replicates our bibliography style.

mjpost commented on bin/anthology/formatter.py:
```diff
@@ -1,6 +1,6 @@
 #
-# Copyright 2019 Marcel Bollmann <marcel@bollmann.me>
+# Copyright 2019-2021 Marcel Bollmann <marcel@bollmann.me>
 #
```
2022?

mjpost commented on bin/anthology/formatter.py:
```diff
@@ -43,6 +45,35 @@
+
+def bibtype_to_csl(bibtype):
+    """Convert a BibTeX entry type to a CSL type.
+
```
Can you expand the definition of CSL here, or [add a link](https://citationstyles.org/)? I'd never heard of it before.

mjpost commented on bin/anthology/formatter.py:
```diff
@@ -43,6 +45,35 @@
+        "article": "article-journal",
+        "inproceedings": "paper-conference",
+        "proceedings": "book",
+    }
```
Hopefully we'll have `misc` soon, too, for things like keynotes.

mjpost commented on bin/anthology/papers.py:
```diff
@@ -1,6 +1,6 @@
 #
-# Copyright 2019 Marcel Bollmann <marcel@bollmann.me>
+# Copyright 2019-2021 Marcel Bollmann <marcel@bollmann.me>
 #
```
2022

mjpost commented on bin/anthology/papers.py:
```diff
@@ -357,6 +371,90 @@ def as_bibtex(self, concise=False):
+        """Return a citation object suitable for CiteProcJSON."""
+        if self._citeproc_json is None:
+            json = {
+                "id": self.bibkey,
```
I see you're just using a dictionary here, but `json` is also a module, and since it might get imported some day, I'd suggest using a different name for the object.

mjpost reviewed (approved):
LGTM. My inclination is to keep using the `citeproc` style as our needs may grow, for example if we add a `misc` type. I've also had requests before to publish APA citation styles, for example.

mbollmann commented:
I refactored a bit, moving the Citeproc stuff exclusively into `formatter`, but the CSL type mapping into `Paper`, because we're likely to need more info than just the bibtype for this in the future (e.g. for keynotes).

I also added a hack to add links to paper titles in Citeproc-generated bibliography strings. The ideal solution would probably be to copy and modify the [ACL .csl file](https://github.com/citation-style-language/styles/blob/master/association-for-computational-linguistics.csl), either by keeping a copy internally or even sending a PR to the CSL repo (the ACL style there was last updated 7 years ago). We could keep this hack for now and I'll open a separate issue for this?

I've also changed around the tabs/headings/labels regarding these bibliography strings a bit, both directly on the paper page and in the "cite" popup. Please check if you like them as they are now or would prefer to change some wordings etc.

mjpost commented:
This looks good to me.
